services/galley/src/Galley/API/LegalHold.hs

```diff
@@ -0,0 +1,244 @@
+                Log.err . Log.msg $ "expected to find a prekey for user: "
+                                 <> toByteString' uid <> " but none was found"
+                throwM internalError
+            Just lstKey -> pure lstKey
+        let clientId = clientIdFromPrekey . unpackLastPrekey $ lastKey
+        pure $ UserLegalHoldStatusResponse status (Just lastKey) (Just clientId)
+
+
+-- | Request to provision a device on the legal hold service for a user
+requestDevice :: UserId ::: TeamId ::: UserId ::: JSON -> Galley Response
+requestDevice (zusr ::: tid ::: uid ::: _) = do
+    assertLegalHoldEnabled tid
+
+    membs <- Data.teamMembers tid
+    void $ permissionCheck zusr ChangeLegalHoldUserSettings membs
+
+    userLHStatus <- fmap (view legalHoldStatus) <$> Data.teamMember tid uid
+    case userLHStatus of
+        Just UserLegalHoldEnabled -> throwM userLegalHoldAlreadyEnabled
+        Just UserLegalHoldPending ->  provisionLHDevice <&> setStatus status204
+        Just UserLegalHoldDisabled -> provisionLHDevice <&> setStatus status201
+        Nothing -> throwM teamMemberNotFound
+  where
+    provisionLHDevice :: Galley Response
+    provisionLHDevice = do
+        (lastPrekey', prekeys) <- requestDeviceFromService
+        -- We don't distinguish the last key here; brig will do so when the device is added
+        LegalHoldData.insertPendingPrekeys uid (unpackLastPrekey lastPrekey' : prekeys)
+        LegalHoldData.setUserLegalHoldStatus tid uid UserLegalHoldPending
+        Client.notifyClientsAboutLegalHoldRequest zusr uid lastPrekey'
+        pure empty
+
+    requestDeviceFromService :: Galley (LastPrekey, [Prekey])
+    requestDeviceFromService = do
+        LegalHoldData.dropPendingPrekeys uid
+        lhDevice <- LHService.requestNewDevice tid uid
+        let NewLegalHoldClient prekeys lastKey = lhDevice
+        return (lastKey, prekeys)
+
+-- | Approve the adding of a Legal Hold device to the user
+-- we don't delete pending prekeys during this flow just in case
+-- it gets interupted. There's really no reason to delete them anyways
+-- since they are replaced if needed when registering new LH devices.
+approveDevice
+    :: UserId ::: TeamId ::: UserId ::: ConnId ::: JsonRequest ApproveLegalHoldForUserRequest ::: JSON
+    -> Galley Response
+approveDevice (zusr ::: tid ::: uid ::: connId ::: req ::: _) = do
+    assertLegalHoldEnabled tid
+
+    unless (zusr == uid) (throwM accessDenied)
+    assertOnTeam uid tid
+    ApproveLegalHoldForUserRequest mPassword <- fromJsonBody req
+    ensureReAuthorised zusr mPassword
+    assertUserLHPending
+
+    mPreKeys <- LegalHoldData.selectPendingPrekeys uid
+    (prekeys, lastPrekey') <- case mPreKeys of
+        Nothing -> do
+            Log.info $ Log.msg @Text "No prekeys found"
+            throwM noLegalHoldDeviceAllocated
+        Just keys -> pure keys
+
+    clientId <- Client.addLegalHoldClientToUser uid connId prekeys lastPrekey'
+    -- TODO: this may be redundant
```

Comment:
This _is_ redundant, please remove @fisx

Reply:
5325b1350534f1069cd07fcb893e78873a54d91c